File: Problems/LeetCode/easy/33-duplicate-zeros.py
class Solution(object):
    def duplicateZeros(self, arr):
        zeros = 0
        i = 0
        l = len(arr)
        while i + zeros < l:
            if arr[i] == 0: zeros += 1
            i += 1
        i -= 1
        while zeros > 0:
            if i + zeros < l: arr[i + zeros] = arr[i]
            if arr[i] == 0:
                zeros -= 1
                arr[i + zeros] = arr[i]
            i -= 1

instance = Solution()

arr1 = [9,0,9,0,6,0,0,0,1,1,6,5,4,4,8,3,0,0,0,1,5,3,0,0,7,2,1,0,2,0,9,1,0,2,0,0,0,0,0,0,0,6,0,0,7,9,0,8,7,0,9,7,1,0,2,0,0,0,0,9,0,0,0,0]
instance.duplicateZeros(arr1)
print("Example 1: Result - " + str(arr1))       
print("Example 1: Result - " + str([9,0,0,9,0,0,6,0,0,0,0,0,0,1,1,6,5,4,4,8,3,0,0,0,0,0,0,1,5,3,0,0,0,0,7,2,1,0,0,2,0,0,9,1,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6,0,0]))       

File: Problems/LeetCode/easy/33-duplicate-zeros.js
/*

Given a fixed length array arr of integers, duplicate each occurrence of zero, shifting the remaining elements to the right.

Note that elements beyond the length of the original array are not written.

Do the above modifications to the input array in place, do not return anything from your function.

Example 1:
    Input: [1,0,2,3,0,4,5,0]
    Output: null
    Explanation: After calling your function, the input array is modified to: [1,0,0,2,3,0,0,4]

Example 2:
    Input: [1,2,3]
    Output: null
    Explanation: After calling your function, the input array is modified to: [1,2,3]

Note:

1 <= arr.length <= 10000
0 <= arr[i] <= 9

*/
let duplicateZeros = function (arr) {
    let i = 0, zero = 0;
    for (i = 0; i + zero < arr.length; i++) if (arr[i] === 0) zero++; 

    for (i -= 1; zero > 0; i--) {
        if (i + zero < arr.length) arr[i + zero] = arr[i];
        if (arr[i] === 0) {
            zero--;
            arr[i + zero] = arr[i];
        }
    }
};

let arr1 = [1, 0, 2, 3, 0, 4, 5, 0];
duplicateZeros(arr1);
console.log('Example 1: Result - ' + arr1 + ' Expected: 1,0,0,2,3,0,0,4');
let arr2 = [1,2,3];
duplicateZeros(arr2);
console.log('Example 2: Result - ' + arr2 + ' Expected: 1,2,3');